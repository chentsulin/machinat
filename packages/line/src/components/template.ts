import { MachinatNode } from '@machinat/core/types';
import { unitSegment, partSegment } from '@machinat/core/renderer';
import { PartSegment } from '@machinat/core/renderer/types';
import { annotateLineComponent } from '../utils';
import { LineComponent } from '../types';

/**
 * @category Props
 */
type ButtonTemplateProps = {
  /**
   * Alternative text. Displayed on devices that do not support template
   * messages. Max character limit: 400
   */
  altText: string;
  /** Image URL (Max character limit: 1,000) */
  thumbnailImageUrl?: string;
  /** Alias of `thumbnailImageUrl`. */
  imageUrl?: string;
  /**
   * Aspect ratio of the image, rectangle: 1.51:1, square: 1:1. Default to
   * `'rectangle'`.
   */
  imageAspectRatio?: 'rectangle' | 'square';
  /**
   * Size of the image. Default to `'cover'`.
   * - cover:   The image fills the entire image area. Parts of the image that
   *            do not fit in the area are not displayed.
   * - contain: The entire image is displayed in the image area. A background is
   *            displayed in the unused areas to the left and right of vertical
   *            images and in the areas above and below horizontal images.
   */
  imageSize?: 'contain' | 'cover';
  /**
   * Background color of the image. Specify a RGB color value. Default: #FFFFFF.
   */
  imageBackgroundColor?: string;
  title?: string;
  /** Message text. */
  text: string;
  /**
   * {@link Action} elements displayed as the buttons at the template. Max 4
   * buttons.
   */
  actions: MachinatNode;
  /**
   * An {@link Action} element to be triggered when image, title or text area is
   * tapped.
   */
  defaultAction?: MachinatNode;
};

/** @internal */
const __ButtonTemplate = async function ButtonTemplate(node, path, render) {
  const {
    actions,
    defaultAction,
    altText,
    thumbnailImageUrl,
    imageUrl,
    imageAspectRatio,
    imageSize,
    imageBackgroundColor,
    title,
    text,
  } = node.props;

  const defaultActionSegments = await render(defaultAction, '.defaultAction');
  const defaultActionValue = defaultActionSegments?.[0].value;

  const actionSegments = await render(actions, '.actions');
  const actionValues = actionSegments?.map((seg) => seg.value);

  return [
    unitSegment(node, path, {
      type: 'template' as const,
      altText,
      template: {
        type: 'buttons',
        thumbnailImageUrl: thumbnailImageUrl || imageUrl,
        imageAspectRatio,
        imageSize,
        imageBackgroundColor,
        title,
        text,
        defaultAction: defaultActionValue,
        actions: actionValues,
      },
    }),
  ];
};
/**
 * Template with an image, title, text, and multiple action buttons.
 * @category Component
 * @props {@link ButtonTemplateProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#buttons).
 */
export const ButtonTemplate: LineComponent<ButtonTemplateProps> = annotateLineComponent(
  __ButtonTemplate
);

/**
 * @category Props
 */
type ConfirmTemplateProps = {
  /**
   * Alternative text. Displayed on devices that do not support template
   * messages. Max character limit: 400
   */
  altText: string;
  /**
   * Exactly 2 {@link Action} elements displayed as the buttons at the template.
   */
  actions: MachinatNode;
  /** Message text. */
  text: string;
};

/** @internal */
const __ConfirmTemplate = async function ConfirmTemplate(node, path, render) {
  const { actions, altText, text } = node.props;
  const actionSegments = await render(actions, '.actions');
  const actionsValues = actionSegments?.map((segment) => segment.value);

  return [
    unitSegment(node, path, {
      type: 'template' as const,
      altText,
      template: {
        type: 'confirm',
        text,
        actions: actionsValues,
      },
    }),
  ];
};
/**
 * Template with two action buttons.
 * @category Component
 * @props {@link ConfirmTemplateProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#confirm).
 */
export const ConfirmTemplate: LineComponent<ConfirmTemplateProps> = annotateLineComponent(
  __ConfirmTemplate
);

/**
 * @category Props
 */
type CarouselItemProps = {
  /**
   * {@link Action} elements displayed as the buttons at the template. Max 3
   * buttons.
   */
  actions: MachinatNode;
  /**
   * An {@link Action} element to be triggered when image, title or text area is
   * tapped.
   */
  defaultAction?: MachinatNode;
  /** Image URL (Max character limit: 1,000) */
  thumbnailImageUrl?: string;
  /** Alias of `thumbnailImageUrl`. */
  imageUrl?: string;
  /**
   * Background color of the image. Specify a RGB color value. The default
   * value is #FFFFFF (white).
   */
  imageBackgroundColor?: string;
  title?: string;
  /** Message text. */
  text?: string;
};

/** @internal */
const __CarouselItem = async function CarouselItem(node, path, render) {
  const {
    actions,
    defaultAction,
    imageUrl,
    thumbnailImageUrl,
    imageBackgroundColor,
    title,
    text,
  } = node.props;

  const defaultActionSegments = await render(defaultAction, '.defaultAction');
  const defaultActionValue = defaultActionSegments?.[0].value;

  const actionSegments = await render(actions, '.actions');
  const actionValues = actionSegments?.map((segment) => segment.value);

  return [
    partSegment(node, path, {
      thumbnailImageUrl: thumbnailImageUrl || imageUrl,
      imageBackgroundColor,
      title,
      text,
      defaultAction: defaultActionValue,
      actions: actionValues,
    }),
  ];
};
/**
 * Column items of {@link CarouselTemplate}.
 * @category Component
 * @props {@link CarouselItemProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#column-object-for-carousel).
 */
export const CarouselItem: LineComponent<
  CarouselItemProps,
  PartSegment<any>
> = annotateLineComponent(__CarouselItem);

/**
 * @category Props
 */
type CarouselTemplateProps = {
  /**
   * {@link CarouselItem} elements displayed as bubble columns in the carousel.
   * Max 10 columns.
   */
  children: MachinatNode;
  /**
   * Alternative text. Displayed on devices that do not support template
   * messages. Max character limit: 400
   */
  altText: string;
  /**
   * Aspect ratio of the image, rectangle: 1.51:1, square: 1:1. Default to
   * `'rectangle'`.
   */
  imageAspectRatio?: 'rectangle' | 'square';
  /**
   * Size of the image. Default to `'cover'`.
   * - cover:   The image fills the entire image area. Parts of the image that
   *            do not fit in the area are not displayed.
   * - contain: The entire image is displayed in the image area. A background is
   *            displayed in the unused areas to the left and right of vertical
   *            images and in the areas above and below horizontal images.
   */
  imageSize?: 'contain' | 'cover';
};

/** @internal */
const __CarouselTemplate = async function CarouselTemplate(node, path, render) {
  const { children, altText, imageAspectRatio, imageSize } = node.props;
  const columnSegments = await render(children, '.children');
  const cloumnValues = columnSegments?.map((segment) => segment.value);

  return [
    unitSegment(node, path, {
      type: 'template' as const,
      altText,
      template: {
        type: 'carousel',
        imageAspectRatio,
        imageSize,
        columns: cloumnValues,
      },
    }),
  ];
};
/**
 * Template with multiple columns which can be cycled like a carousel. The
 * columns are shown in order when scrolling horizontally.
 * @category Component
 * @props {@link CarouselTemplateProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#carousel).
 */
export const CarouselTemplate: LineComponent<CarouselTemplateProps> = annotateLineComponent(
  __CarouselTemplate
);

/**
 * @category Props
 */
type ImageCarouselItemProps = {
  /** Image URL (Max character limit: 1,000) */
  imageUrl?: string;
  /** Alias of `imageUrl`. Either one of `url` and `imageUrl` must be specified. */
  url?: string;
  /** One {@link Action} element to be triggered the when image is tapped. */
  action?: MachinatNode;
};

/** @internal */
const __ImageCarouselItem = async function ImageCarouselItem(
  node,
  path,
  render
) {
  const { url, imageUrl, action } = node.props;
  const actionSegments = await render(action, '.action');
  const actionValue = actionSegments?.[0].value;

  return [
    partSegment(node, path, {
      imageUrl: imageUrl || url,
      action: actionValue,
    }),
  ];
};
/**
 * Column items of {@link ImageCarouselTemplate}.
 * @category Component
 * @props {@link ImageCarouselItemProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#column-object-for-image-carousel).
 */
export const ImageCarouselItem: LineComponent<
  ImageCarouselItemProps,
  PartSegment<any>
> = annotateLineComponent(__ImageCarouselItem);

/**
 * @category Props
 */
type ImageCarouselTemplateProps = {
  /** {@link ImageCarouselItem} elements displayed as image columns in the
   * carousel. Max 10 columns.
   */
  children: MachinatNode;
  /**
   * Alternative text. Displayed on devices that do not support template
   * messages. Max character limit: 400
   */
  altText: string;
};

/** @internal */
const __ImageCarouselTemplate = async function ImageCarouselTemplate(
  node,
  path,
  render
) {
  const { children, altText } = node.props;
  const columnSegments = await render(children, '.children');
  const columnValues = columnSegments?.map((segment) => segment.value);

  return [
    unitSegment(node, path, {
      type: 'template' as const,
      altText,
      template: {
        type: 'image_carousel',
        columns: columnValues,
      },
    }),
  ];
};
/**
 * Template with multiple images which can be cycled like a carousel. The images
 * are shown in order when scrolling horizontally.
 * @category Component
 * @props {@link ImageCarouselTemplateProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#image-carousel).
 */
export const ImageCarouselTemplate: LineComponent<ImageCarouselTemplateProps> = annotateLineComponent(
  __ImageCarouselTemplate
);
