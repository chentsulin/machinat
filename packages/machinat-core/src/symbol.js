// @flow
export const MACHINAT_ELEMENT_TYPE = Symbol('machinat.element');

export const MACHINAT_FRAGMENT_TYPE = Symbol('machinat.element.fragment');

export const MACHINAT_PAUSE_TYPE = Symbol('machinat.element.pause');

export const MACHINAT_PROVIDER_TYPE = Symbol('machinat.element.provider');

export const MACHINAT_THUNK_TYPE = Symbol('machinat.element.thunk');

export const MACHINAT_RAW_TYPE = Symbol('machinat.element.raw');

export const MACHINAT_NATIVE_TYPE = Symbol('machinat.component.native');

export const MACHINAT_SERVICES_CONTAINER = Symbol(
  'machinat.services.container'
);

export const MACHINAT_SERVICES_INTERFACE = Symbol(
  'machinat.services.interface'
);

export const MACHINAT_SERVICES_PROVIDER = Symbol('machinat.services.provider');
