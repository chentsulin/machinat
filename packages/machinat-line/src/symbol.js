// @flow
export const LINE_NAITVE_TYPE = Symbol.for('machinat.line.native');

export const NO_RENDERED = Symbol.for('machinat.line.no_rendered');
