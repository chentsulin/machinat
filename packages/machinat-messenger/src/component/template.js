import invariant from 'invariant';
import { joinTextualSegments, valuesOfAssertedType } from 'machinat-utility';

import * as buttonModule from './button';
import { asMessagesUnitComponent, asPartComponent } from './utils';

const buttonComponents = Object.values(buttonModule);

const CHILDREN = '.children';

const getButtonValues = valuesOfAssertedType(...buttonComponents);
const getUrlButtonValues = valuesOfAssertedType(buttonModule.URLButton);

const GenericItem = (
  {
    props: {
      children,
      title,
      imageURL,
      subtitle,
      defaultAction: defaultActionProp,
    },
  },
  render
) => {
  const buttonSegments = render(children, CHILDREN);
  const defaultActionSegments = render(defaultActionProp, '.defaultAction');

  let defaultAction;
  if (defaultActionSegments !== null) {
    const defaultActionValues = getUrlButtonValues(defaultActionSegments);

    // ignore title field
    const [{ title: _, ...defaultActionVal }] = defaultActionValues;
    defaultAction = defaultActionVal;
  }

  return [
    {
      title,
      image_url: imageURL,
      subtitle,
      default_action: defaultAction,
      buttons: getButtonValues(buttonSegments),
    },
  ];
};
const __GenericItem = asPartComponent(GenericItem);

const getGenericItemValues = valuesOfAssertedType(__GenericItem);

const GenericTemplate = (
  { props: { children, sharable, imageAspectRatio } },
  render
) => [
  {
    message: {
      attachment: {
        type: 'template',
        payload: {
          template_type: 'generic',
          sharable,
          image_aspect_ratio: imageAspectRatio,
          elements: getGenericItemValues(render(children, CHILDREN)),
        },
      },
    },
  },
];
const __GenericTemplate = asMessagesUnitComponent(GenericTemplate);

const ListTemplate = (
  { props: { children, topStyle, sharable, button } },
  render
) => [
  {
    message: {
      attachment: {
        type: 'template',
        payload: {
          template_type: 'list',
          top_element_style: topStyle,
          sharable,
          elements: getGenericItemValues(render(children, CHILDREN)),
          buttons: getButtonValues(render(button, '.button')),
        },
      },
    },
  },
];
const __ListTemplate = asMessagesUnitComponent(ListTemplate);

const ButtonTemplate = ({ props: { children, text, sharable } }, render) => {
  const segments = joinTextualSegments(render(text, '.text'));

  let textValue;
  invariant(
    segments !== null &&
      segments.length === 1 &&
      typeof (textValue = segments[0].value) === 'string', // eslint-disable-line prefer-destructuring
    segments
      ? `<br /> in prop "text" of <ButtonTemplate /> is invalid`
      : `prop "text" of <ButtonTemplate /> should not be empty`
  );

  return [
    {
      message: {
        attachment: {
          type: 'template',
          payload: {
            template_type: 'button',
            text: textValue,
            sharable,
            buttons: getButtonValues(render(children, CHILDREN)),
          },
        },
      },
    },
  ];
};
const __ButtonTemplate = asMessagesUnitComponent(ButtonTemplate);

const MediaTemplate = (
  { props: { children, type, attachmentId, url, sharable } },
  render
) => [
  {
    message: {
      attachment: {
        type: 'template',
        payload: {
          template_type: 'media',
          sharable,
          elements: [
            {
              media_type: type,
              url,
              attachment_id: attachmentId,
              buttons: getButtonValues(render(children, CHILDREN)),
            },
          ],
        },
      },
    },
  },
];
const __MediaTemplate = asMessagesUnitComponent(MediaTemplate);

const OpenGraphTemplate = ({ props: { children, url, sharable } }, render) => [
  {
    message: {
      attachment: {
        type: 'template',
        payload: {
          template_type: 'open_graph',
          sharable,
          elements: [
            {
              url,
              buttons: getButtonValues(render(children, CHILDREN)),
            },
          ],
        },
      },
    },
  },
];
const __OpenGraphTemplate = asMessagesUnitComponent(OpenGraphTemplate);

const ReceiptItem = ({
  props: { title, subtitle, quantity, price, currency, imageURL },
}) => [
  {
    title,
    subtitle,
    quantity,
    price,
    currency,
    image_url: imageURL,
  },
];
const __ReceiptItem = asPartComponent(ReceiptItem);

const getReceiptItemValues = valuesOfAssertedType(__ReceiptItem);

const ReceiptTemplate = (
  {
    props: {
      children,
      sharable,
      recipientName,
      orderNumber,
      currency,
      paymentMethod,
      orderURL,
      timestamp,
      address,
      summary,
      adjustments,
    },
  },
  render
) => [
  {
    message: {
      attachment: {
        type: 'template',
        payload: {
          template_type: 'receipt',
          sharable,
          recipient_name: recipientName,
          order_number: orderNumber,
          currency,
          payment_method: paymentMethod,
          order_url: orderURL,
          timestamp:
            timestamp instanceof Date
              ? `${Math.floor(timestamp.getTime() / 1000)}`
              : timestamp,
          address,
          summary,
          adjustments,
          elements: getReceiptItemValues(render(children, CHILDREN)),
        },
      },
    },
  },
];

const __ReceiptTemplate = asMessagesUnitComponent(ReceiptTemplate);

export {
  __GenericItem as GenericItem,
  __GenericTemplate as GenericTemplate,
  __ListTemplate as ListTemplate,
  __ButtonTemplate as ButtonTemplate,
  __MediaTemplate as MediaTemplate,
  __OpenGraphTemplate as OpenGraphTemplate,
  __ReceiptItem as ReceiptItem,
  __ReceiptTemplate as ReceiptTemplate,
};
