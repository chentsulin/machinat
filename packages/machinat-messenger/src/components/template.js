import formatNode from '@machinat/core/utils/formatNode';
import { unitSegment, partSegment } from '@machinat/core/renderer';
import { annotateMessengerComponent } from '../utils';

export const GenericItem = async (node, path, render) => {
  const {
    buttons,
    title,
    imageURL,
    subtitle,
    defaultAction: defaultActionProp,
  } = node.props;

  const buttonSegments = await render(buttons, '.buttons');
  const buttonValues = buttonSegments?.map((segment) => segment.value);

  const defaultActionSegments = await render(
    defaultActionProp,
    '.defaultAction'
  );

  let defaultAction;
  if (defaultActionSegments !== null) {
    const { title: _, ...restOfURLButton } = defaultActionSegments[0].value;
    defaultAction = restOfURLButton;
  }

  return [
    partSegment(node, path, {
      title,
      image_url: imageURL,
      subtitle,
      default_action: defaultAction,
      buttons: buttonValues,
    }),
  ];
};
annotateMessengerComponent(GenericItem);

export const GenericTemplate = async (node, path, render) => {
  const { children, sharable, imageAspectRatio } = node.props;
  const elementsSegments = await render(children, '.children');
  const elementValues = elementsSegments?.map((segment) => segment.value);

  return [
    unitSegment(node, path, {
      message: {
        attachment: {
          type: 'template',
          payload: {
            template_type: 'generic',
            sharable,
            image_aspect_ratio: imageAspectRatio,
            elements: elementValues,
          },
        },
      },
    }),
  ];
};
annotateMessengerComponent(GenericTemplate);

export const ButtonTemplate = async (node, path, render) => {
  const { children, buttons, sharable } = node.props;
  const textSegments = await render(children, '.children');
  let text = '';

  if (textSegments) {
    for (const segment of textSegments) {
      if (segment.type !== 'text') {
        throw new TypeError(
          `non-textual node ${formatNode(
            segment.node
          )} received, only textual nodes allowed`
        );
      }
    }

    text = textSegments[0].value;
  }

  const buttonSegments = await render(buttons, '.buttons');
  const buttonValues = buttonSegments?.map((segment) => segment.value);

  return [
    unitSegment(node, path, {
      message: {
        attachment: {
          type: 'template',
          payload: {
            template_type: 'button',
            text,
            sharable,
            buttons: buttonValues,
          },
        },
      },
    }),
  ];
};
annotateMessengerComponent(ButtonTemplate);

export const MediaTemplate = async (node, path, render) => {
  const { buttons, type, attachmentId, url, sharable } = node.props;
  const buttonSegments = await render(buttons, '.buttons');
  const buttonValues = buttonSegments?.map((segment) => segment.value);

  return [
    unitSegment(node, path, {
      message: {
        attachment: {
          type: 'template',
          payload: {
            template_type: 'media',
            sharable,
            elements: [
              {
                media_type: type,
                url,
                attachment_id: attachmentId,
                buttons: buttonValues,
              },
            ],
          },
        },
      },
    }),
  ];
};
annotateMessengerComponent(MediaTemplate);

export const OpenGraphTemplate = async (node, path, render) => {
  const { buttons, url, sharable } = node.props;
  const buttonSegments = await render(buttons, '.buttons');
  const buttonValues = buttonSegments?.map((segment) => segment.value);

  return [
    unitSegment(node, path, {
      message: {
        attachment: {
          type: 'template',
          payload: {
            template_type: 'open_graph',
            sharable,
            elements: [
              {
                url,
                buttons: buttonValues,
              },
            ],
          },
        },
      },
    }),
  ];
};
annotateMessengerComponent(OpenGraphTemplate);

export const ReceiptItem = async (node, path) => {
  const { title, subtitle, quantity, price, currency, imageURL } = node.props;
  return [
    partSegment(node, path, {
      title,
      subtitle,
      quantity,
      price,
      currency,
      image_url: imageURL,
    }),
  ];
};
annotateMessengerComponent(ReceiptItem);

export const ReceiptTemplate = async (node, path, render) => {
  const {
    children,
    sharable,
    recipientName,
    merchantName,
    orderNumber,
    currency,
    paymentMethod,
    orderURL,
    timestamp,
    address,
    summary,
    adjustments,
  } = node.props;

  const elementSegments = await render(children, '.children');
  const elementValues = elementSegments?.map((segment) => segment.value);

  return [
    unitSegment(node, path, {
      message: {
        attachment: {
          type: 'template',
          payload: {
            template_type: 'receipt',
            sharable,
            recipient_name: recipientName,
            merchant_name: merchantName,
            order_number: orderNumber,
            currency,
            payment_method: paymentMethod,
            order_url: orderURL,
            timestamp:
              timestamp instanceof Date
                ? `${Math.floor(timestamp.getTime() / 1000)}`
                : timestamp,
            address,
            summary,
            adjustments,
            elements: elementValues,
          },
        },
      },
    }),
  ];
};
annotateMessengerComponent(ReceiptTemplate);
