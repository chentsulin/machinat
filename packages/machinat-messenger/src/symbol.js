export const MESSENGER_NAITVE_TYPE = Symbol.for('machinat.messenger.native');
