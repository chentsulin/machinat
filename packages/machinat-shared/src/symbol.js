export const MACHINAT_ELEMENT_TYPE = Symbol.for('machinat.element');

export const MACHINAT_ROOT_ELEMENT = Symbol.for('machinat.element.root');

export const MACHINAT_FRAGMENT_TYPE = Symbol.for('machinat.element.fragment');

export const MACHINAT_ASYNC_TYPE = Symbol.for('machinat.element.async');
