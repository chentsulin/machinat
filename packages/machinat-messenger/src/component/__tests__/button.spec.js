import Machinat from 'machinat';

import { MESSENGER_NAITVE_TYPE } from '../../symbol';
import {
  URLButton,
  PostbackButton,
  ShareButton,
  BuyButton,
  CallButton,
  LoginButton,
  LogoutButton,
  GamePlayButton,
} from '../button';
import { GenericTemplate, GenericItem } from '../template';
import renderHelper from './renderHelper';

const renderInside = jest.fn();
const render = renderHelper(renderInside);
beforeEach(() => {
  renderInside.mockReset();
});

test.each([
  URLButton,
  PostbackButton,
  ShareButton,
  BuyButton,
  CallButton,
  LoginButton,
  LogoutButton,
  GamePlayButton,
])('is valid Component', Button => {
  expect(typeof Button).toBe('function');
  expect(Button.$$native).toBe(MESSENGER_NAITVE_TYPE);
  expect(Button.$$entry).toBe(undefined);
  expect(Button.$$unit).toBe(false);
});

describe('URLButton', () => {
  it('match snapshot', () => {
    expect(
      [
        <URLButton title="my button" url="http://machinat.com" />,
        <URLButton
          title="my button"
          url="http://machinat.com"
          heightRatio="compact"
          extensions
          fallbackURL="http://..."
          hideShareButton
        />,
      ].map(render)
    ).toMatchSnapshot();
  });
});

describe('PostbackButton', () => {
  it('match snapshot', () => {
    expect(
      render(<PostbackButton title="my button" payload="_MY_PAYLOAD_" />)
    ).toMatchSnapshot();
  });
});

describe('ShareButton', () => {
  it('match snapshot', () => {
    const sharedTemplate = (
      <GenericTemplate>
        <GenericItem title="foo" subtitle="bar" />
      </GenericTemplate>
    );
    renderInside.mockImplementation(node =>
      node
        ? [
            {
              value: {
                message: '__RENDERED_GENERIC_TEMPLATE_MEASSGE_OBJ__',
              },
              node: sharedTemplate,
            },
          ]
        : null
    );
    expect(
      [<ShareButton />, <ShareButton>{sharedTemplate}</ShareButton>].map(render)
    ).toMatchSnapshot();

    expect(renderInside).toHaveBeenCalledWith(sharedTemplate, '.children');
  });

  it('throw if non GenericTemplate children given', () => {
    const Invalid = () => {};
    renderInside.mockImplementation(node => [
      {
        value: '__SOMETHING_WRONG__',
        node,
      },
    ]);

    expect(() =>
      render(
        <ShareButton>
          <Invalid />
        </ShareButton>
      )
    ).toThrowErrorMatchingInlineSnapshot(
      `"<Invalid /> is invalid in .children, only <[GenericTemplate]/> allowed"`
    );
  });
});

describe('BuyButton', () => {
  it('match snapshot', () => {
    expect(
      render(
        <BuyButton
          title="my button"
          payload="_MY_PAYLOAD_"
          currency="USD"
          isTest
          paymentType="FIXED_AMOUNT"
          merchantName="My Fake Business"
          requestedInfo={[
            'shipping_address',
            'contact_name',
            'contact_phone',
            'contact_email',
          ]}
          priceList={[
            {
              label: 'subtotal',
              amount: '12.75',
            },
          ]}
        />
      )
    ).toMatchSnapshot();
  });
});

describe('CallButton', () => {
  it('match snapshot', () => {
    expect(
      render(<CallButton title="call me maybe" number="+15105551234" />)
    ).toMatchSnapshot();
  });
});

describe('LoginButton', () => {
  it('match snapshot', () => {
    expect(
      render(<LoginButton url="https://council.elrond" />)
    ).toMatchSnapshot();
  });
});

describe('LoginButton', () => {
  it('match snapshot', () => {
    expect(
      render(<LoginButton url="https://council.elrond" />)
    ).toMatchSnapshot();
  });
});

describe('LogoutButton', () => {
  it('match snapshot', () => {
    expect(render(<LogoutButton />)).toMatchSnapshot();
  });
});

describe('GamePlayButton', () => {
  it('match snapshot', () => {
    expect(
      render(
        <GamePlayButton
          title="I want to play a game"
          payload="GAME_OVER"
          playerId="Adam"
          contextId="SAW"
        />
      )
    ).toMatchSnapshot();
  });
});
