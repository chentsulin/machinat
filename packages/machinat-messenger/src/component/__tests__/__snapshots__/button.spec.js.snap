// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuyButton match snapshot 1`] = `
Array [
  Object {
    "node": Object {
      "$$typeof": Symbol(machinat.element),
      "props": Object {
        "currency": "USD",
        "isTest": true,
        "merchantName": "My Fake Business",
        "payload": "_MY_PAYLOAD_",
        "paymentType": "FIXED_AMOUNT",
        "priceList": Array [
          Object {
            "amount": "12.75",
            "label": "subtotal",
          },
        ],
        "requestedInfo": Array [
          "shipping_address",
          "contact_name",
          "contact_phone",
          "contact_email",
        ],
        "title": "my button",
      },
      "type": [Function],
    },
    "path": "$",
    "type": "part",
    "value": Object {
      "payload": "_MY_PAYLOAD_",
      "payment_summary": Object {
        "currency": "USD",
        "is_test_payment": true,
        "merchant_name": "My Fake Business",
        "payment_type": "FIXED_AMOUNT",
        "price_list": Array [
          Object {
            "amount": "12.75",
            "label": "subtotal",
          },
        ],
        "requested_user_info": Array [
          "shipping_address",
          "contact_name",
          "contact_phone",
          "contact_email",
        ],
      },
      "title": "my button",
      "type": "payment",
    },
  },
]
`;

exports[`CallButton match snapshot 1`] = `
Array [
  Object {
    "node": Object {
      "$$typeof": Symbol(machinat.element),
      "props": Object {
        "number": "+15105551234",
        "title": "call me maybe",
      },
      "type": [Function],
    },
    "path": "$",
    "type": "part",
    "value": Object {
      "number": "+15105551234",
      "title": "call me maybe",
      "type": "phone_number",
    },
  },
]
`;

exports[`GamePlayButton match snapshot 1`] = `
Array [
  Object {
    "node": Object {
      "$$typeof": Symbol(machinat.element),
      "props": Object {
        "contextId": "SAW",
        "payload": "GAME_OVER",
        "playerId": "Adam",
        "title": "I want to play a game",
      },
      "type": [Function],
    },
    "path": "$",
    "type": "part",
    "value": Object {
      "game_metadata": Object {
        "context_id": "SAW",
        "player_id": "Adam",
      },
      "payload": "GAME_OVER",
      "title": "I want to play a game",
      "type": "game_play",
    },
  },
]
`;

exports[`LoginButton match snapshot 1`] = `
Array [
  Object {
    "node": Object {
      "$$typeof": Symbol(machinat.element),
      "props": Object {
        "url": "https://council.elrond",
      },
      "type": [Function],
    },
    "path": "$",
    "type": "part",
    "value": Object {
      "type": "account_link",
      "url": "https://council.elrond",
    },
  },
]
`;

exports[`LoginButton match snapshot 2`] = `
Array [
  Object {
    "node": Object {
      "$$typeof": Symbol(machinat.element),
      "props": Object {
        "url": "https://council.elrond",
      },
      "type": [Function],
    },
    "path": "$",
    "type": "part",
    "value": Object {
      "type": "account_link",
      "url": "https://council.elrond",
    },
  },
]
`;

exports[`LogoutButton match snapshot 1`] = `
Array [
  Object {
    "node": Object {
      "$$typeof": Symbol(machinat.element),
      "props": Object {},
      "type": [Function],
    },
    "path": "$",
    "type": "part",
    "value": Object {
      "type": "account_unlink",
    },
  },
]
`;

exports[`PostbackButton match snapshot 1`] = `
Array [
  Object {
    "node": Object {
      "$$typeof": Symbol(machinat.element),
      "props": Object {
        "payload": "_MY_PAYLOAD_",
        "title": "my button",
      },
      "type": [Function],
    },
    "path": "$",
    "type": "part",
    "value": Object {
      "payload": "_MY_PAYLOAD_",
      "title": "my button",
      "type": "postback",
    },
  },
]
`;

exports[`ShareButton match snapshot 1`] = `
Array [
  Array [
    Object {
      "node": Object {
        "$$typeof": Symbol(machinat.element),
        "props": Object {},
        "type": [Function],
      },
      "path": "$",
      "type": "part",
      "value": Object {
        "share_contents": undefined,
        "type": "element_share",
      },
    },
  ],
  Array [
    Object {
      "node": Object {
        "$$typeof": Symbol(machinat.element),
        "props": Object {
          "children": Object {
            "$$typeof": Symbol(machinat.element),
            "props": Object {
              "children": Object {
                "$$typeof": Symbol(machinat.element),
                "props": Object {
                  "subtitle": "bar",
                  "title": "foo",
                },
                "type": [Function],
              },
            },
            "type": [Function],
          },
        },
        "type": [Function],
      },
      "path": "$",
      "type": "part",
      "value": Object {
        "share_contents": "__RENDERED_GENERIC_TEMPLATE_MEASSGE_OBJ__",
        "type": "element_share",
      },
    },
  ],
]
`;

exports[`URLButton match snapshot 1`] = `
Array [
  Array [
    Object {
      "node": Object {
        "$$typeof": Symbol(machinat.element),
        "props": Object {
          "title": "my button",
          "url": "http://machinat.com",
        },
        "type": [Function],
      },
      "path": "$",
      "type": "part",
      "value": Object {
        "fallback_url": undefined,
        "messenger_extensions": undefined,
        "title": "my button",
        "type": "web_url",
        "url": "http://machinat.com",
        "webview_height_ratio": undefined,
        "webview_share_button": undefined,
      },
    },
  ],
  Array [
    Object {
      "node": Object {
        "$$typeof": Symbol(machinat.element),
        "props": Object {
          "extensions": true,
          "fallbackURL": "http://...",
          "heightRatio": "compact",
          "hideShareButton": true,
          "title": "my button",
          "url": "http://machinat.com",
        },
        "type": [Function],
      },
      "path": "$",
      "type": "part",
      "value": Object {
        "fallback_url": "http://...",
        "messenger_extensions": true,
        "title": "my button",
        "type": "web_url",
        "url": "http://machinat.com",
        "webview_height_ratio": "compact",
        "webview_share_button": "hide",
      },
    },
  ],
]
`;
